src/main/java/com/epam/aidial/core/controller/AccessControlBaseController.java

```diff
@@ -21,37 +22,66 @@ public abstract class AccessControlBaseController {
 
     /**
      * @param bucket url encoded bucket name
-     * @param path url encoded resource path
+     * @param path   url encoded resource path
      */
     public Future<?> handle(String resourceType, String bucket, String path) {
         ResourceType type = ResourceType.of(resourceType);
         String urlDecodedBucket = UrlUtil.decodePath(bucket);
         String decryptedBucket = proxy.getEncryptionService().decrypt(urlDecodedBucket);
-        boolean hasReadAccess = isSharedWithMe(type, bucket, path);
-        boolean hasWriteAccess = hasWriteAccess(path, decryptedBucket);
-        boolean hasAccess = checkFullAccess ? hasWriteAccess : hasReadAccess || hasWriteAccess;
-
-        if (!hasAccess) {
-            return context.respond(HttpStatus.FORBIDDEN, "You don't have an access to the bucket " + bucket);
+        if (decryptedBucket == null) {
+            context.respond(HttpStatus.FORBIDDEN, "You don't have an access to the bucket " + bucket);
+            return Future.succeededFuture();
         }
 
+        // we should take a real user bucket not provided from resource
+        String actualUserLocation = BlobStorageUtil.buildInitiatorBucket(context);
+        String actualUserBucket = proxy.getEncryptionService().encrypt(actualUserLocation);
+
         ResourceDescription resource;
         try {
             resource = ResourceDescription.fromEncoded(type, urlDecodedBucket, decryptedBucket, path);
         } catch (Exception ex) {
             String errorMessage = ex.getMessage() != null ? ex.getMessage() : DEFAULT_RESOURCE_ERROR_MESSAGE.formatted(path);
-            return context.respond(HttpStatus.BAD_REQUEST, errorMessage);
+            context.respond(HttpStatus.BAD_REQUEST, errorMessage);
+            return Future.succeededFuture();
         }
 
-        handle(resource);
-        return Future.succeededFuture();
+        Future<Void> permissionFuture = proxy.getVertx().executeBlocking(() -> {
+            boolean hasReadAccess = isSharedResource(resource, actualUserBucket, actualUserLocation);
+            boolean hasWriteAccess = hasWriteAccess(path, decryptedBucket);
+            boolean hasAccess = checkFullAccess ? hasWriteAccess : hasReadAccess || hasWriteAccess;
+
+            if (!hasAccess) {
+                throw new HttpException(HttpStatus.FORBIDDEN, "You don't have an access to the bucket " + bucket);
```

Comment:
Does it make sense to say that you do not have access to the specific resource instead of bucket?